### PR TITLE
Refactor the UI of the verification flow

### LINKED ISSUE #313: continuous loading indicator is shown on Second & third fields when user auto paste attestations.
**Frequency:** 100%
**Repro on build version:** Android internal build V1.13.1(1004294349) 
**Repro on devices:** One plus 6T (10),Samsung Galaxy Note 5(7) , Redmi Note 8 (9.0)
**Pre-Condition:** Make user codes should paste automatically
   **Repro Steps:**

1. Launch the application 
2. Complete on boarding education
3. Tap on create new account 
4. Enter pin
5. Enter phone number 
6. Tap on start button 
7. Wait for phone number confirmation fields 
8. Make sure codes should auto paste 
9. Observe continuous loading indicator is shown on Second & third fields 

**Current Behavior:**  Continuous loading indicator is shown 
**Expected Behavior:**  Codes should auto paste & user should redirect to connect contacts screen. 
**Observation:** Codes are not auto pasted in ios. 
**Impact:** User  blocked to confirm phone number . 
**Attachment:**  [ContinuousLoadingIndicator.mp4](https://drive.google.com/file/d/1MmfuDGEXoMju7xSb8gFAvazWE-kwBFWG/view?usp=sharing)

### PR DESCRIPTION
**Pull request by gnardini**
### Description

The UI of the verification flow was very disconnected from the saga. There was virtually no error handling, so any wrong codes made the UI get stuck completely.

This PR adds a better connection between the UI and the saga, especially around error handling. Most of the relevant changed code is in `packages/mobile/src/verify/VerificationCodeInput.tsx`.

Something important that is missing from this PR is unit tests. For now this has been tested manually extensively, but more manual testing is still needed and also some automated tests.

### Other changes

N/A

### Tested

Manually on Android and iOS, trying all the edge cases that i could think of. I could have certainly missed some so any help testing this is greatly appreciated.

### How others should test

Go through the verification flow, try any border case you can think of and make sure you never get stuck and can always verify.

### Related issues

- Fixes #313

### Backwards compatibility

N/A

### CODE REVIEW

File: packages/mobile/src/verify/VerificationCodeInput.test.tsx [hunk 1]
```diff
@@ -1,59 +1,47 @@
...
-        inputPlaceholder="Test placeholder"
-        inputPlaceholderWithClipboardContent="Test clipboard"
-        onInputChange={jest.fn()}
-        isCodeSubmitting={true}
-        attestationCodes={[]}
-        numCompleteAttestations={0}
-      />
-    )
-    expect(toJSON()).toMatchSnapshot()
+// TODO: Add better tests for this
```
Comment (jeanregisser) [inline packages/mobile/src/verify/VerificationCodeInput.test.tsx:10]: Indeed, these snapshot tests don't seem very helpful.

File: packages/mobile/src/verify/VerificationInputScreen.test.tsx [hunk 2]
```diff
@@ -6,6 +6,7 @@ import { Screens } from 'src/navigator/Screens'
 import VerificationInputScreen from 'src/verify/VerificationInputScreen'
 import { createMockStore, getMockStackScreenProps } from 'test/utils'
 
+// TODO: Add better tests for this
```
Comment (jeanregisser) [inline packages/mobile/src/verify/VerificationInputScreen.test.tsx:9]: 👍

File: packages/mobile/src/verify/VerificationCodeInput.tsx [hunk 3]
```diff
@@ -4,55 +4,138 @@ import {
...
   label: string
   index: number // index of code in attestationCodes array
   inputValue: string // the raw code inputed by the user
   inputPlaceholder: string
   inputPlaceholderWithClipboardContent: string
-  isCodeSubmitting: boolean // is the inputted code being processed
-  onInputChange: (value: string) => void
-  attestationCodes: AttestationCode[] // The codes in the redux store
-  numCompleteAttestations: number // has the code been accepted and completed
+  onInputChange: (value: string, runValidations?: boolean) => void
```
Comment (jeanregisser) [inline packages/mobile/src/verify/VerificationCodeInput.tsx:28]: What does `runValidations` mean?

**Review by jeanregisser — APPROVED** [review]
Great!! Thanks for fixing this so quickly.

File: packages/mobile/src/components/CodeInput.test.tsx [hunk 4]
```diff
@@ -6,11 +6,10 @@ import CodeInput, { CodeInputStatus } from 'src/components/CodeInput'
 describe('CodeInput', () => {
   it('renders correctly for all CodeInputStatus states', () => {
     ;[
-      CodeInputStatus.DISABLED,
-      CodeInputStatus.INPUTTING,
-      CodeInputStatus.PROCESSING,
-      CodeInputStatus.RECEIVED,
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/components/CodeInput.test.tsx:12]: I had removed the state before re-adding it but forgot to add it again in the test. Added `Error` state as well

File: packages/mobile/src/identity/actions.ts [hunk 5]
```diff
@@ -482,3 +494,12 @@ export const reportRevealStatus = (
     pepper,
   }
 }
+
+export const setAttestationInputStatus = (
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/identity/actions.ts:498]: I agree, this was just the simplest way to fix this without making larger changes. In the future we should add a bigger refactor of this that includes fixing the saga-UI relationship which will involve having better redux actions

File: packages/mobile/src/verify/utils.ts [hunk 6]
```diff
@@ -24,3 +25,23 @@ export function getPhoneNumberState(
     }
   }
 }
+
+export function isCodeRepeated(codes: string[], value: string) {
+  return codes.reduce((count, item) => count + (item === value ? 1 : 0), 0) >= 2
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/verify/utils.ts:30]: We are looking for more than one instance of the value, so I don't think `indexOf` does the trick. I agree it's not very clear though, it started simpler and grew to this, I'll just add a for loop which will be easier to read

File: packages/mobile/src/verify/VerificationCodeInput.tsx (hunk 3 above)
Comment (gnardini) [reply packages/mobile/src/verify/VerificationCodeInput.tsx:28]: Renamed to `processCodeIfValid`. This is because if a user already has some completed attestations we want to show them in the UI but not process them/send them via komenci again.

It's a bit convoluted to be honest, right now the logic for all the UI stuff is in the individual inputs `VerificationCodeInput`. In a future refactor, part of this logic could be moved to the saga and part to the `VerificationInputScreen`. I didn't add it in the `VerificationInputScreen` right now even though I think that's a better place because it's still a class component and I didn't want to either have to change it to a function component or remember how all the lifecycle events worked with class components. I've been trying to avoid them as much as possible lately.

File: packages/mobile/src/verify/utils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeanregisser) [reply packages/mobile/src/verify/utils.ts]: I see, I thought this was checking if the given value was already in the array. 👍

**Comment by Celoqa** [comment]
Hi @gnardini @jeanregisser We have verified this issue on latest Android internal build V1.14.2 & Testflight build V1.14.2 and observe the following
**Android:**

- [ ] User is able to complete the verification flow by auto pasting the codes 
- [ ] Not able to see continuous loading indicator & without any edge case failing 

**iOS:**

- [ ] User not able to Auto paste the codes when user try to complete the verification flow 
 Can you please let us know it is expected behavior in iOS [Attachment](https://drive.google.com/file/d/1k4sYSOvAeitZvrw2tnzay5q4SIQ6C5fP/view?usp=sharing)

**Comment by gnardini** [comment]
Hi @Celoqa 

Thanks for testing. Auto-pasting is only supported on Android currently so codes need to be pasted manually on iOS.

**Comment by Celoqa** [comment]
Thanks @gnardini for giving clarification.
